README.md
# esni-rev-proxy
Golang reverse proxy with support of ESNI 01-draft (https://tools.ietf.org/html/draft-ietf-tls-esni-01) on top of TLS 1.3

__Motovation:__ As of April 2020 ESNI, is still a draft extension for TLS 1.3 and not officialy supported by OpenSSL and major projects like nginx, apache, etc (however unofficial forks exists). This project porvides a tiny golang reverse proxy that can terminate TLS 1.3 wint ESNI and forward plain HTTP to upstream. This covers the gap, and if you want to experiment with ESNI draft you can use it right now with your vanilla nginx/apache etc. Also your browser should support ESNI, and for now it is only firefox (https://www.elliotjreed.com/post/security/2019-07-08_Enable_DNS_over_HTTPS_and_Encrypted_SNI_in_Firefox). 

This project is extension of tris-localserver example code : https://github.com/cloudflare/tls-tris/tree/pwu/esni-consolidated/_dev/tris-localserver and was inspired by discussion: https://serverfault.com/questions/976377/how-can-i-set-up-encrypted-sni-on-my-own-servers

---
__How to build:__
As declared here https://github.com/cloudflare/tls-tris/tree/pwu/esni-consolidated, since crypto/tls is very deeply (and not that elegantly) coupled with the Go stdlib, it is impossible to vendor it as crypto/tls because stdlib packages would import the standard one and mismatch. 
Approach here would be to build custom GOROOT (that has patched standard libraries), and then on top of it, build the current code.

Build ___esni reverse proxy___:
1. `git clone https://github.com/devopsext/esni-rev-proxy.git`
2. `cd esni-rev-proxy && git checkout v1.0.1` 
3. `prepareGoRoot.sh` - this script create patched GOROOT folder (`.GOROOT/`) in current directory (applicable to Linux/MacOS)
4. `export GOROOT=$(pwd)/.GOROOT`
5. `go mod vendor`
6. `go build`

Build ___esnitool___ (in case use need to generate esni keys):
Esnitool source code is copied from cloudflare repo: 
https://github.com/cloudflare/tls-tris/tree/pwu/esni-consolidated/_dev/esnitool
1. Copy `GOROOT` folder, that was prepared by script (see steps above) into esnitool/: `cp -r GOROOT/* esnitool/`
2. `cd esnitool/ && export GOROOT=$(pwd)/GOROOT`
3. `go build`


__How to run:__
1. Generate esni keys pair (public and private) with `esnitool`:
Just follow inline help of esnitool:
```
Usage of ./esnitool:
  -esni-keys-file string
        Write base64-encoded ESNI keys to file instead of stdout
  -esni-private-file string
        Write ESNI private key to file instead of stdout
  -validity duration
        Validity period of the keys (default 24h0m0s)
```
As example:
`./esnitool  -esni-keys-file ./esni.pub -esni-private-file ./esni -validity 32h`
Pay attention to validity period. Specify as long as you need. In case you set it short,
you need to rotate key pair.

2. Setup specific DNS record for your server (in addition to A or AAAA record), that you are going to access with ESNI:
 Create TXT record for host name `_esni.<YOUR-DOMAIN>` with the contents of `esni.pub` file (esni public key)
 
3. Run `esni-reverse-proxy` in front of your web server, to accept and terminate TLS 1.3 with ESNI,
and forward decrypted traffic to your web-server. `esni-rev-proxy` has self explanatory flags:
```
Usage of ./esni-rev-proxy:
  -b string
        Address:port used for binding (default "0.0.0.0:443")
  -cert value
        Triplet of SNI:PrivateKey.File:CertChain.File
  -cliauth
        Performs client authentication (RequireAndVerifyClientCert used)
  -esni-keys string
        File with base64-encoded ESNIKeys
  -esni-private string
        Private key file for ESNI
  -pq string
        Enable quantum-resistant algorithms [c: Support classical and Quantum-Resistant, q: Enable Quantum-Resistant only]
  -rtt0 string
        0-RTT, accepts following values [n: None, a: Accept, o: Offer, oa: Offer and Accept] (default "n")
  -showaccesslog
        Show access log
  -upstream string
        Upstream URL to forward traffic to
  -stats-metrics-bind string
        Address:port used for binding. Metrics available at /metrics (prometheus format),
        health-check at /helathz (default "0.0.0.0:8181")
```
Short comments:
* `-cert` flag used to set SNI name and corresponding cert/key pair, in case your proxy severe several domains, you can specify several values.
* `-upstream` is an upstream to forward plain traffic to.

Example:

```
esni-rev-proxy -b 0.0.0.0:443 \
-esni-keys esni.pub -esni-private esni \
-cert "www.example.com:/mycerts/www.example.com.key:/mycerts/www.example.com.crt" \
-cert "other-domain.com:/mycerts/other-domain.com.key:/mycerts/other-domain.com.crt" \
-upstream http://internal-endpoint \
-showaccesslog
```

This would start up the reverse proxy that:
1. Accept incoming connections on all interfaces on port 443
2. Decrypt ESNI (using `esni.pub` and `esni` public and private key pair) and choose appropriate certificate for host name `www.example.com` or `other-domain.com`,
stored in `/mycerts`. The key/cert pair specified in first `-cert` flag is also treated as default cert/key pair - 
so it will be used if no match to SNI host detected.
3. Forward decrypted traffic to `http://internal-endpoint`
4. Print incoming and forwarded requests in stdout (`-showaccesslog` flag)
5. Export 2 endpoints `0.0.0.0:8181/metrics` - metrics in prometheus format and
`0.0.0.0:8181/healthz` - simple http health-check.

---
__Sample list of metrics in prometheus format:__
```
# HELP esnirevproxy_http_average_last_min_rps Incoming HTTP rps average for the last minute
# TYPE esnirevproxy_http_average_last_min_rps gauge
esnirevproxy_http_average_last_min_rps 0

# HELP esnirevproxy_http_upstream_latency_msec Upstream latency in milliseconds
# TYPE esnirevproxy_http_upstream_latency_msec histogram
# Bucket distribution deleted...
esnirevproxy_http_upstream_latency_msec_sum{upstream="www.google.com"} 14243
esnirevproxy_http_upstream_latency_msec_count{upstream="www.google.com"} 20

# HELP esnirevproxy_http_upstream_response_codes Upstream response HTTP codes
# TYPE esnirevproxy_http_upstream_response_codes counter
esnirevproxy_http_upstream_response_codes{code="200",upstream="www.google.com"} 7
esnirevproxy_http_upstream_response_codes{code="204",upstream="www.google.com"} 13

# HELP esnirevproxy_tcp_connections_total Total active/idle connections
# TYPE esnirevproxy_tcp_connections_total gauge
esnirevproxy_tcp_connections_total 1

# HELP esnirevproxy_tls_failed_handshakes Total number of failed TLS handshakes
# TYPE esnirevproxy_tls_failed_handshakes counter
esnirevproxy_tls_failed_handshakes 1

# HELP esnirevproxy_tls_handshake_duration_msec Handshake time in milliseconds
# TYPE esnirevproxy_tls_handshake_duration_msec histogram
# Bucket distribution deleted...
esnirevproxy_tls_handshake_duration_msec_sum 3
esnirevproxy_tls_handshake_duration_msec_count 1

# HELP esnirevproxy_tls_successful_handshakes Total number of successful TLS handshakes
# TYPE esnirevproxy_tls_successful_handshakes counter
esnirevproxy_tls_successful_handshakes{sni="localhost"} 1

#Standard go_ metrics is not printed here.
```


CHANGELOG.md
## v1.0.2 [2020-05-08]
#### Release Notes
Various improvements based on load testing.
Load-testing was performed by wrk, here the results:
```bash
wrk -t50 -c1000 -d360s 'https://127.0.0.1' --timeout 15s
Running 6m test @ https://127.0.0.1
  50 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.45s   743.48ms  13.88s    80.11%
    Req/Sec    17.64     13.36   120.00     75.77%
  254094 requests in 6.00m, 7.48GB read
Requests/sec:    705.57
Transfer/sec:     21.27MB 
```
#### Features
- Reworked scripts for preparing GOROOT (now accepts 1 input that point out the target folder for prepared GOROOT)
- Added `build.sh` - script that prepare GOROOT and build binary.
- Added script `enable_esni_firefox_darwin.sh` - this is example of script to automate ESNI support enablement in fire fox browser
- Add CPU and MEM profiling (-cpuprof, -memprof flags)
#### Bugfixes

- Fix data race when gathering tls handshake statistics.

## v1.0.1 [2020-04-09]
#### Features
- Added telemetry in prometheus format

## v1.0.0 [2020-04-03]
#### Release Notes
Initial version

esnirevproxy.go
package main

import (
	"context"
	"crypto/tls" //This is patched library
	"crypto/x509"
	"encoding/base64"
	"flag"
	"github.com/prometheus/client_golang/prometheus"
	"io/ioutil"
	"log"
	"net"
	"net/http" //This is patched library
	"net/http/httputil"
	"net/url"
	"os"
	"os/signal"
	"reflect"
	"runtime"
	"runtime/pprof"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/paulbellamy/ratecounter"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

type ZeroRTT_t int

// Bitset
const (
	ZeroRTT_None   ZeroRTT_t = 0
	ZeroRTT_Offer            = 1 << 0
	ZeroRTT_Accept           = 1 << 1
)

type revProxy struct {
	upstreamURL  *url.URL
	reverseProxy *httputil.ReverseProxy
}

const (
	tlsNewConn = iota
	tlsActiveConn
	tlsClosedConn
)

type tlsConnEvent struct {
	remoteAddr string
	connID     []byte
	sni        string
	evType     uint
	when       time.Time
}

type rpTransport struct {
	http.RoundTripper
}

func (t *rpTransport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
	start := time.Now()
	resp, err = t.RoundTripper.RoundTrip(req)
	if req != nil && resp != nil {
		upstreamHTTPResponseCodesCounterVec.WithLabelValues(req.Host, strconv.Itoa(resp.StatusCode)).Inc()
		upstreamLatencyCounterVec.WithLabelValues(req.Host).Observe(float64(time.Now().Sub(start).Milliseconds()))
	}
	return resp, err
}

type esniRevProxy struct {
	httpServer *http.Server
	wgServer   sync.WaitGroup

	httpStatsServer *http.Server
	wgStatServer    sync.WaitGroup

	rp            *revProxy
	showAccessLog bool

	//stats
	wg               sync.WaitGroup
	counter          *ratecounter.RateCounter // RPS counting
	rpsTickerDone    chan bool
	tslConnEventChan chan tlsConnEvent
}

var tlsVersionToName = map[uint16]string{
	tls.VersionTLS10: "1.0",
	tls.VersionTLS11: "1.1",
	tls.VersionTLS12: "1.2",
	tls.VersionTLS13: "1.3",
}

//Prometheus metrics
var (
	failedTLSHandshakesCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Namespace: "esnirevproxy",
			Subsystem: "tls",
			Name:      "failed_handshakes",
			Help:      "Total number of failed TLS handshakes",
		},
	)
	successfulTLSHandshakesCounterVec = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: "esnirevproxy",
			Subsystem: "tls",
			Name:      "successful_handshakes",
			Help:      "Total number of successful TLS handshakes",
		},
		// Host label (aka SNI)
		[]string{"sni"},
	)

	tlsHandshakeDurationHist = prometheus.NewHistogram(
		prometheus.HistogramOpts{
			Namespace: "esnirevproxy",
			Subsystem: "tls",
			Buckets:   []float64{5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000},
			Name:      "handshake_duration_msec",
			Help:      "Handshake time in milliseconds"})

	totalConnectionsGauge = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Namespace: "esnirevproxy",
			Subsystem: "tcp",
			Name:      "connections_total",
			Help:      "Total active/idle connections",
		},
	)

	rpsGauge = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Namespace: "esnirevproxy",
			Subsystem: "http",
			Name:      "average_last_min_rps",
			Help:      "Incoming HTTP rps average for the last minute"})

	upstreamHTTPResponseCodesCounterVec = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: "esnirevproxy",
			Subsystem: "http",
			Name:      "upstream_response_codes",
			Help:      "Upstream response HTTP codes",
		},
		[]string{"upstream", "code"},
	)

	upstreamLatencyCounterVec = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Namespace: "esnirevproxy",
			Subsystem: "http",
			Buckets:   []float64{100, 500, 1000, 2500, 5000, 10000},
			Name:      "upstream_latency_msec",
			Help:      "Upstream latency in milliseconds"},
		[]string{"upstream"},
	)
)

type arrayFlags []string

func (i *arrayFlags) String() string {
	if i == nil {
		return ""
	}
	return strings.Join(*i, ",")

}

func (i *arrayFlags) Set(value string) error {
	*i = append(*i, value)
	return nil
}

func buildTLSConig(sniCertsChain *arrayFlags,
	argZeroRTT string,
	clientAuth bool,
	pq string,
	esniPrivateKeyFile, esniKeysFile string) *tls.Config {

	var (
		err            error
		zeroRTT        = ZeroRTT_None
		esniKeys       *tls.ESNIKeys
		esniPrivateKey []byte
	)

	tlsConfig := tls.Config{
		MaxVersion:        tls.VersionTLS13,
		ClientAuth:        tls.NoClientCert,
		NameToCertificate: map[string]*tls.Certificate{},
	}

	//SNI -> key,cert
	for i, elem := range *sniCertsChain {
		//Splitting to SNI, Key and Cert:
		arr := strings.Split(elem, ":")
		if len(arr) != 3 {
			log.Fatalf("Corrupted value '%s', shpold be: SNI:PrivateKey.File:CertChain.File", elem)
		}
		if i == 0 { //Default cert
			log.Printf("Default esniRevProxy cert/key pair would be sourced from %s", elem)
		}

		keyStr, err := ioutil.ReadFile(arr[1])
		if err != nil {
			log.Fatalf("Can't read from file '%s': %v", arr[1], err)
		}
		certStr, err := ioutil.ReadFile(arr[2])
		if err != nil {
			log.Fatalf("Can't read from file '%s': %v", arr[2], err)
		}

		cert, err := tls.X509KeyPair(certStr, keyStr)
		if err != nil {
			log.Fatalf("Can't parse X509 key pair from key:cert '%s:%s': %v", arr[1], arr[2], err)
		}
		tlsConfig.Certificates = append(tlsConfig.Certificates, cert)
		tlsConfig.NameToCertificate[arr[0]] = &cert

	}

	switch argZeroRTT {
	case "a":
		zeroRTT = ZeroRTT_Accept
	case "o":
		zeroRTT = ZeroRTT_Offer
	case "oa":
		zeroRTT = ZeroRTT_Offer | ZeroRTT_Accept
	}

	if (zeroRTT & ZeroRTT_Offer) == ZeroRTT_Offer {
		tlsConfig.Max0RTTDataSize = 100 * 1024
	}

	if clientAuth {
		tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
	}
	tlsConfig.ClientCAs = x509.NewCertPool()
	tlsConfig.Accept0RTTData = (zeroRTT & ZeroRTT_Accept) == ZeroRTT_Accept

	tlsConfig.NextProtos = []string{"npn_proto"}

	if pq != "" {
		tlsConfig.CurvePreferences = append(tlsConfig.CurvePreferences, []tls.CurveID{tls.HybridSIDHp503Curve25519, tls.HybridSIKEp503Curve25519}...)
		if pq == "c" {
			tlsConfig.CurvePreferences = append(tlsConfig.CurvePreferences, []tls.CurveID{tls.X25519, tls.CurveP256, tls.CurveP384, tls.CurveP521}...)
		}
	}

	//ESNI keys
	if esniPrivateKeyFile != "" {
		esniPrivateKey, err = ioutil.ReadFile(esniPrivateKeyFile)
		if err != nil {
			log.Fatalf("Failed to read ESNI private key: %s", err)
		}
	}
	if esniKeysFile != "" {
		contents, err := ioutil.ReadFile(esniKeysFile)
		if err != nil {
			log.Fatalf("Failed to read ESNIKeys: %s", err)
		}
		esniKeysBytes, err := base64.StdEncoding.DecodeString(string(contents))
		if err != nil {
			log.Fatalf("Bad -esni-keys: %s", err)
		}
		esniKeys, err = tls.ParseESNIKeys(esniKeysBytes)
		if esniKeys == nil {
			log.Fatalf("Cannot parse ESNIKeys: %s", err)
		}
		tlsConfig.GetServerESNIKeys = func([]byte) (*tls.ESNIKeys, []byte, error) { return esniKeys, esniPrivateKey, nil }
	}

	if keyLogFile := os.Getenv("SSLKEYLOGFILE"); keyLogFile != "" {
		tlsConfig.KeyLogWriter, err = os.OpenFile(keyLogFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
		if err != nil {
			log.Fatalf("Cannot open keylog file: %v", err)
		}
		log.Println("Keylog enabled...")
	}
	return &tlsConfig
}

func newESNIRevProxy(
	address string,
	sniCertsChain *arrayFlags,
	argZeroRTT string,
	clientAuth bool,
	pq string,
	esniPrivateKeyFile, esniKeysFile string,
	upstream string,
	showAccessLog bool,
	addrStats string) *esniRevProxy {

	var (
		err         error
		upstreamURL *url.URL
	)
	upstreamURL, err = url.Parse(upstream)
	if err != nil {
		log.Fatalf("Can't parse upstream URL: %v", err)
	}

	rp := revProxy{
		upstreamURL:  upstreamURL,
		reverseProxy: httputil.NewSingleHostReverseProxy(upstreamURL),
	}
	//Set round tripper to get upstream statistics
	rp.reverseProxy.Transport = &rpTransport{http.DefaultTransport}

	s := new(esniRevProxy)

	mux := http.NewServeMux()
	mux.Handle("/", http.HandlerFunc(s.requestHandler))
	s.httpServer = &http.Server{
		Addr:      address,
		Handler:   mux,
		ConnState: s.gatherTLSConnStats,
		TLSConfig: buildTLSConig(sniCertsChain, argZeroRTT, clientAuth, pq, esniPrivateKeyFile, esniKeysFile)}

	muxStats := http.NewServeMux()
	muxStats.Handle("/metrics", promhttp.Handler())
	muxStats.Handle("/healthz", http.HandlerFunc(s.healthzHandler))
	s.httpStatsServer = &http.Server{
		Addr:    addrStats,
		Handler: muxStats,
	}
	s.showAccessLog = showAccessLog
	s.rp = &rp
	s.counter = ratecounter.NewRateCounter(60 * time.Second)
	return s
}

func (s *esniRevProxy) start() {
	log.Println("Starting esni reverse proxy...")
	//Primary
	s.wgServer.Add(1)
	s.tslConnEventChan = make(chan tlsConnEvent)
	go func() {
		defer s.wgServer.Done()
		if err := s.httpServer.ListenAndServeTLS("", ""); err != http.ErrServerClosed {
			log.Fatalf("ListenAndServeTLS(): %v", err)
		}
	}()

	//Stats & healthz
	s.wgStatServer.Add(1)
	go func() {
		defer s.wgStatServer.Done()
		if err := s.httpStatsServer.ListenAndServe(); err != http.ErrServerClosed {
			log.Fatalf("ListenAndServe(): %v", err)
		}
	}()

	//rps counter
	s.wg.Add(1)
	s.rpsTickerDone = make(chan bool)
	go func() {
		rpsTicker := time.NewTicker(time.Second)
		defer rpsTicker.Stop()
		defer s.wg.Done()

		for {
			select {
			case <-s.rpsTickerDone:
				return
			case <-rpsTicker.C:
				rpsGauge.Set(float64(s.counter.Rate()) / 60)
			}
		}
	}()

	//tls handshake duration counter
	s.wg.Add(1)
	go func(connEvents <-chan tlsConnEvent) {
		defer s.wg.Done()

		connectionStats := map[string]map[string]interface{}{}

		for event := range connEvents {
			remoteAddr := event.remoteAddr
			if remoteAddr == "" {
				continue
			}
			switch event.evType {
			case tlsNewConn:
				connectionStats[remoteAddr] = map[string]interface{}{}
				connectionStats[remoteAddr]["startedAt"] = event.when
			case tlsActiveConn:
				if _, ok := connectionStats[remoteAddr]["tlsConnectionID"]; !ok {
					successfulTLSHandshakesCounterVec.WithLabelValues(event.sni).Inc()
					tlsHandshakeDurationHist.Observe(
						float64(
							event.when.
								Sub(connectionStats[remoteAddr]["startedAt"].(time.Time)).Milliseconds()))

					connectionStats[remoteAddr]["tlsConnectionID"] = event.connID
				}
			case tlsClosedConn:
				delete(connectionStats, remoteAddr)
			}
		}
	}(s.tslConnEventChan)

}

func (s *esniRevProxy) stop() {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := s.httpServer.Shutdown(ctx); err != nil {
		log.Printf("Can't graceful shutdown server: %v", err)
		s.httpServer.Close()
	}
	s.wgServer.Wait()

	if err := s.httpStatsServer.Shutdown(ctx); err != nil {
		log.Printf("Can't graceful shutdown healthz/metrics server: %v", err)
		s.httpStatsServer.Close()
	}
	s.wgStatServer.Wait()

	//Stopping rps counter
	close(s.rpsTickerDone)

	//Stopping tls handshake duration counter
	close(s.tslConnEventChan)

	s.wg.Wait()
}

func (s *esniRevProxy) requestHandler(w http.ResponseWriter, r *http.Request) {

	//TODO: move this counter to middleware
	s.counter.Incr(1)

	if s.showAccessLog {
		log.Printf("==>INPUT REQUEST \"%v %s\", SNI:%s\n[headers]: %v\n", r.Method, r.URL, r.TLS.ServerName, r.Header)
	}
	// Update the headers to allow for SSL redirection
	r.URL.Host = s.rp.upstreamURL.Host
	r.URL.Scheme = s.rp.upstreamURL.Scheme
	if r.Header.Get("Host") == "" && r.TLS.ServerName != "" {
		r.Header.Set("X-Forwarded-Host", r.TLS.ServerName)
		r.Header.Set("Host", r.TLS.ServerName)
	} else if r.Header.Get("Host") == "" && r.TLS.ServerName == "" {
		r.Header.Set("Host", r.Header.Get("Http_host"))
	}
	s.rp.upstreamURL.Port()
	if s.showAccessLog {
		log.Printf("==>FORWARDED REQUEST upstream: %s://%s\n[headers]: %v\n", s.rp.upstreamURL.Scheme, s.rp.upstreamURL.Host, r.Header)
	}
	r.Host = s.rp.upstreamURL.Host

	//We can inject into request the context with call backs to get statistics https://blog.golang.org/http-tracing
	s.rp.reverseProxy.ServeHTTP(w, r)

}

func (s *esniRevProxy) healthzHandler(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
}

func (s *esniRevProxy) gatherTLSConnStats(c net.Conn, state http.ConnState) {
	tlsConn := reflect.ValueOf(c).Interface().(*tls.Conn)
	remoteAddr := tlsConn.RemoteAddr().String()

	switch state {
	case http.StateNew:
		s.tslConnEventChan <- tlsConnEvent{
			remoteAddr: remoteAddr,
			evType:     tlsNewConn,
			when:       time.Now(),
		}
		totalConnectionsGauge.Inc()
	case http.StateActive:
		if tlsConn.ConnectionState().HandshakeComplete {
			sni := tlsConn.ConnectionState().ServerName
			if sni == "" {
				sni = reflect.ValueOf(tlsConn.RemoteAddr()).Interface().(*net.TCPAddr).IP.String()
			}
			s.tslConnEventChan <- tlsConnEvent{
				remoteAddr: remoteAddr,
				connID:     tlsConn.ConnectionState().ConnectionID,
				sni:        sni,
				evType:     tlsActiveConn,
				when:       time.Now(),
			}
		}

	case http.StateClosed:
		s.tslConnEventChan <- tlsConnEvent{
			remoteAddr: remoteAddr,
			evType:     tlsClosedConn,
		}
		totalConnectionsGauge.Dec()

		//connection closed without established TLS handshake,
		//this means that TLS handshake failed
		if !tlsConn.ConnectionState().HandshakeComplete {
			failedTLSHandshakesCounter.Inc()
		}

	}

}

func main() {
	var arg_sniCertsChain arrayFlags

	arg_addr := flag.String("b", "0.0.0.0:443", "Address:port used for binding")
	flag.Var(&arg_sniCertsChain, "cert", "Triplet of SNI:PrivateKey.File:CertChain.File")
	arg_zerortt := flag.String("rtt0", "n", `0-RTT, accepts following values [n: None, a: Accept, o: Offer, oa: Offer and Accept]`)
	//arg_confirm := flag.Bool("rtt0ack", false, "0-RTT confirm")
	arg_clientauth := flag.Bool("cliauth", false, "Performs client authentication (RequireAndVerifyClientCert used)")
	arg_pq := flag.String("pq", "", "Enable quantum-resistant algorithms [c: Support classical and Quantum-Resistant, q: Enable Quantum-Resistant only]")
	arg_esniKeys := flag.String("esni-keys", "", "File with base64-encoded ESNIKeys")
	arg_esniPrivate := flag.String("esni-private", "", "Private key file for ESNI")
	arg_upstream := flag.String("upstream", "", "Upstream URL to forward traffic to")
	arg_accesslog := flag.Bool("showaccesslog", false, "Show access log")
	arg_addrStats := flag.String("stats-metrics-bind", "0.0.0.0:8181", "Address:port used for binding. Metrics available at /metrics (prometheus format), health-check at /helathz")

	arg_cpuProfile := flag.String("cpuprof", "", "CPU profile output file")
	arg_memProfile := flag.String("memprof", "", "Memory profile output file")

	flag.Parse()

	if *arg_cpuProfile != "" {
		f, err := os.Create(*arg_cpuProfile)
		if err != nil {
			log.Fatal("could not create CPU profile: ", err)
		}
		defer f.Close()
		if err := pprof.StartCPUProfile(f); err != nil {
			log.Fatal("could not start CPU profile: ", err)
		}
		defer pprof.StopCPUProfile()
	}

	if *arg_esniPrivate == "" && *arg_esniKeys != "" ||
		*arg_esniPrivate != "" && *arg_esniKeys == "" {
		log.Fatal("Both -esni-keys and -esni-private must be provided.")
	}

	if *arg_upstream == "" {
		log.Fatal("Upstream URL is not set")
	}

	s := newESNIRevProxy(
		*arg_addr,
		&arg_sniCertsChain,
		*arg_zerortt,
		*arg_clientauth,
		*arg_pq,
		*arg_esniPrivate, *arg_esniKeys,
		*arg_upstream,
		*arg_accesslog,
		*arg_addrStats)

	s.start()

	term := make(chan os.Signal)
	signal.Notify(term, os.Interrupt, syscall.SIGTERM)

	select {
	case <-term:
		log.Println("Received SIGTERM, exiting gracefully...")
		s.stop()
	}

	if *arg_memProfile != "" {
		f, err := os.Create(*arg_memProfile)
		if err != nil {
			log.Fatal("could not create memory profile: ", err)
		}
		defer f.Close()
		runtime.GC() // get up-to-date statistics
		if err := pprof.WriteHeapProfile(f); err != nil {
			log.Fatal("could not write memory profile: ", err)
		}
	}

}

func init() {
	prometheus.MustRegister(failedTLSHandshakesCounter)
	prometheus.MustRegister(successfulTLSHandshakesCounterVec)
	prometheus.MustRegister(totalConnectionsGauge)
	prometheus.MustRegister(tlsHandshakeDurationHist)
	prometheus.MustRegister(rpsGauge)
	prometheus.MustRegister(upstreamHTTPResponseCodesCounterVec)
	prometheus.MustRegister(upstreamLatencyCounterVec)
}
